src/main/java/net/unicon/lti/security/lti/LTI3OAuthProviderProcessingFilter.java

```diff
@@ -70,59 +69,20 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
-            List<String> ltiState = (List<String>) httpServletRequest.getSession().getAttribute("lti_state");
-            if (!ltiState.contains(state)) {
-                throw new IllegalStateException("LTI request doesn't contains the expected state");
-            }
-            //Third, we validate the state to be sure that is correct
-            Jws<Claims> stateClaims = ltijwtService.validateState(state);
 
-            // Once we have the state validated we need the key to check the JWT signature from the id_token,
-            // and extract all the values in the LTI3Request object.
-            // Most of the platforms will provide a JWK repo URL and we will have it stored in configuration,
-            // where they store the public keys
-            // With that URL and the "kid" in the header of the jwt id_token, we can get the public key too.
-            // In our tool we have included a alternative mechanism for those platforms without JWK endpoint
-            // The state provides us the way to find that key in our repo. This is not a requirement in LTI, it is just a way to do it that we've implemented, but each one can use the
-            // state in a different way.
-            String jwt = httpServletRequest.getParameter("id_token");
-            if (StringUtils.hasText(jwt)) {
-                //Now we validate the JWT token
-                Jws<Claims> jws = ltijwtService.validateJWT(jwt, stateClaims.getBody().getAudience());
-                if (jws != null) {
-                    //Here we create and populate the LTI3Request object and we will add it to the httpServletRequest, so the redirect endpoint will have all that information
-                    //ready and will be able to use it.
-                    LTI3Request lti3Request = new LTI3Request(httpServletRequest, ltiDataService, true, link, null); // IllegalStateException if invalid
-                    httpServletRequest.setAttribute("LTI3", true); // indicate this request is an LTI3 one
-                    httpServletRequest.setAttribute("lti3_valid", lti3Request.isLoaded() && lti3Request.isComplete()); // is LTI3 request totally valid and complete
-                    httpServletRequest.setAttribute("lti3_message_type", lti3Request.getLtiMessageType()); // is LTI3 request totally valid and complete
-                    httpServletRequest.setAttribute(LTI3Request.class.getName(), lti3Request); // make the LTI3 data accessible later in the request if needed
+            NonceState nonceState = ltiDataService.getRepos().nonceStateRepository.findByStateHash(state);
+            if (nonceState != null) {
+                Jws<Claims> stateClaims = ltijwtService.validateState(nonceState.getState());
+                String nonce = stateClaims.getBody().getId();
+                if (!nonceState.getNonce().equals(nonce)) {
+                    throw new IllegalStateException("LTI request doesn't contains the expected state/nonce");
```

Comment:
Could you change this message to "LTI request doesn't contain the expected state/nonce"?

Reply:
Damn s... always there!